lib/Dialect/Torch/Transforms/AdjustCallingConventions.cpp

```diff
@@ -187,8 +188,49 @@ class AdjustCallingConventionForReturn
 };
 } // namespace
 
+static bool isValidNonContainerResultType(Type result_type) {
+  if (result_type.isa<Torch::BaseTensorType>() ||
+      result_type.isa<Torch::FloatType>() ||
+      result_type.isa<Torch::IntType>() ||
+      result_type.isa<Torch::BoolType>()) {
+    return true;
+  }
+  return false;
+}
+
+static LogicalResult validateReturns(func::FuncOp func) {
+  // getResultTypes can only ever be one item.  Multiple return values are
+  // stored in a tuple.
+  const auto& result_type = func.getResultTypes().front();
+
+  // Allow single tensor returns
+  if (isValidNonContainerResultType(result_type)) {
+    return success();
+  }
+
+  // Allow multi-tensor tuple returns
+  if (auto tuple = result_type.dyn_cast<Torch::TupleType>()) {
+    bool contains_only_tensors = true;
+    const auto& contained_types = tuple.getContainedTypes();
+    for (const auto& contained_type : contained_types) {
+      if (!isValidNonContainerResultType(contained_type)) {
+        contains_only_tensors = false;
+      }
+    }
+    if (contained_types.size() >= 2 && contains_only_tensors) {
+      return success();
+    }
+  }
+
+  return func->emitError(
+    "Functions must return a tensor, multiple tensors, "
+    "or a tuple of multiple tensors.");
```

Comment:
This error message is not completely accurate. Returning values other than tensors is supported. Also, I think we can improve the specificity of the error messages by returning early once an invalid return type is encountered. In other words, rewriting things as:

```cpp
static LogicalResult validateReturns(func::FuncOp func) {
  // ...
  if (auto tuple = resultType.dyn_cast<Torch::TupleType>()) {
    // ...
    if (containedTypes.size() < 2)
      return func->emitError("Error about tuple size");
    // ...
    if (!containsValidTypes)
      return func->emitError("Error about tuple not containing valid types");
  }

  if (!isValidNonContainerResultType(resultType)) {
    return func->emitError("Error about resultType not being a valid type or tuple of valid types");
  }

  return success();
}
```

Reply:
@silvasean's preference was an allowlist rather than a denylist, so that's what I went with here.  Will see what he thinks.

Reply:
Yes, I think an allowlist is better here, since we really should know exactly everything. The errors should be great nonetheless.

Reply:
Resolving.  Updated error message, though I still don't like it a ton.  If you see a potential improvement please let me know.